core/model/src/main/java/org/eclipse/rdf4j/model/util/GraphComparisons.java

```diff
@@ -0,0 +1,415 @@
+			Map<BNode, HashCode> hashDoublePrime = hashBNodes(m, clonedHash);
+
+			Multimap<HashCode, BNode> partitionPrime = partitionMapping(hashDoublePrime);
+			if (isFine(partitionPrime)) {
+				finePartitionMappings.add(hashDoublePrime);
+				if (lowestFound == null || mappingSize(hashDoublePrime).compareTo(mappingSize(blankNodeMapping)) < 0) {
+					lowestFound = hashDoublePrime;
+				}
+			} else {
+				Map<BNode, BNode> compatibleAutomorphism = findCompatibleAutomorphism(fixpoints, finePartitionMappings);
+				if (compatibleAutomorphism != null) {
+					// prune
+					continue;
+				}
+				lowestFound = distinguish(m, hashDoublePrime, partitionPrime, lowestFound, fixpoints,
+						finePartitionMappings);
+			}
+		}
+
+		return lowestFound;
+	}
+
+	protected static Map<BNode, BNode> findCompatibleAutomorphism(List<BNode> fixpoints,
+			List<Map<BNode, HashCode>> partitionMappings) {
+		// check if two mappings with identical hash codes exist
+		for (Map<BNode, HashCode> mapping : partitionMappings) {
+			Map<BNode, HashCode> compatibleMapping = null;
+			for (Map<BNode, HashCode> om : partitionMappings) {
+				if (om.equals(mapping)) {
+					continue;
+				}
+
+				List<HashCode> difference = new ArrayList<>(om.values());
+				difference.removeAll(mapping.values());
+				if (difference.isEmpty()) {
+					compatibleMapping = om;
+					break;
+				}
+			}
+
+			if (compatibleMapping != null) {
+				Map<HashCode, BNode> invertedMapping = mapping.entrySet()
+						.stream()
+						.collect(Collectors.toMap(Entry::getValue, Entry::getKey));
+
+				Map<BNode, BNode> automorphism = new HashMap<>();
+				for (Entry<BNode, HashCode> entry : compatibleMapping.entrySet()) {
+					automorphism.put(entry.getKey(), invertedMapping.get(entry.getValue()));
+				}
+				// check if fixpoints all map, if so we have a compatible automorphism
+				for (BNode fixpoint : fixpoints) {
+					if (!automorphism.get(fixpoint).equals(fixpoint)) {
+						break;
+					}
+					return automorphism;
+				}
+			}
+		}
+		return null;
+	}
+
+	protected static List<Collection<BNode>> partitions(Multimap<HashCode, BNode> partitionMapping) {
+		List<Collection<BNode>> partition = new ArrayList<>();
+
+		for (Entry<HashCode, Collection<BNode>> entry : partitionMapping.asMap().entrySet()) {
+			partition.add(entry.getValue());
+		}
+		return partition;
+	}
+
+	protected static Multimap<HashCode, BNode> partitionMapping(Map<BNode, HashCode> blankNodeMapping) {
+		return Multimaps.invertFrom(Multimaps.forMap(blankNodeMapping), HashMultimap.create());
+	}
+
+	private static BigInteger mappingSize(Map<BNode, HashCode> mapping) {
+		BigInteger size = mapping.values()
+				.stream()
+				.map(hashCode -> new BigInteger(1, hashCode.asBytes()))
+				.reduce(BigInteger.ZERO, (v1, v2) -> v1.add(v2));
+		return size;
+	}
+
+	private static Model labelModel(Model original, Map<BNode, HashCode> hash) {
+		Model result = new LinkedHashModel(original.size());
```

Comment:
Worth benchmarking against the dynamic model. Kinda depends what you end up doing with the model.

Reply:
Note that this model creation only happens once, right at the end, so I suspect the performance difference will not be major - but yeah we can try.